substrate/client/src/light.rs

```diff
@@ -0,0 +1,189 @@
+	is_best: bool,
+}
+
+impl backend::Backend for Backend {
+	type BlockImportOperation = BlockImportOperation;
+	type Blockchain = Blockchain;
+	type State = OnDemandState;
+
+	fn begin_operation(&self, _block: BlockId) -> error::Result<Self::BlockImportOperation> {
+		Ok(BlockImportOperation {
+			pending_block: None,
+		})
+	}
+
+	fn commit_operation(&self, operation: Self::BlockImportOperation) -> error::Result<()> {
+		if let Some(pending_block) = operation.pending_block {
+			let hash = header_hash(&pending_block.header);
+			self.blockchain.storage.insert(hash, pending_block.header, pending_block.justification, None, pending_block.is_best);
+		}
+		Ok(())
+	}
+
+	fn blockchain(&self) -> &Blockchain {
+		&self.blockchain
+	}
+
+	fn state_at(&self, block: BlockId) -> error::Result<Self::State> {
+		Ok(OnDemandState {
+			_block: self.blockchain.storage.id(block).ok_or(error::ErrorKind::UnknownBlock(block))?,
+		})
+	}
+}
+
+impl backend::RemoteBackend for Backend {}
+
+impl backend::BlockImportOperation for BlockImportOperation {
+	type State = OnDemandState;
+
+	fn state(&self) -> error::Result<Option<&Self::State>> {
+		// None means 'locally-stateless' backend
+		Ok(None)
+	}
+
+	fn set_block_data(&mut self, header: block::Header, _body: Option<block::Body>, justification: Option<primitives::bft::Justification>, is_new_best: bool) -> error::Result<()> {
+		assert!(self.pending_block.is_none(), "Only one block per operation is allowed");
+		self.pending_block = Some(PendingBlock {
+			header,
+			justification,
+			is_best: is_new_best,
+		});
+		Ok(())
+	}
+
+	fn update_storage(&mut self, _update: <Self::State as StateBackend>::Transaction) -> error::Result<()> {
+		// we're not storing anything locally => ignore changes
+		Ok(())
+	}
+
+	fn reset_storage<I: Iterator<Item=(Vec<u8>, Vec<u8>)>>(&mut self, _iter: I) -> error::Result<()> {
+		// we're not storing anything locally => ignore changes
+		Ok(())
+	}
+}
+
+impl blockchain::Backend for Blockchain {
+	fn header(&self, id: BlockId) -> error::Result<Option<block::Header>> {
+		self.storage.header(id)
+	}
+
+	fn body(&self, _id: BlockId) -> error::Result<Option<block::Body>> {
+		// TODO [light]: fetch from remote node
+		Ok(None)
+	}
+
+	fn justification(&self, id: BlockId) -> error::Result<Option<primitives::bft::Justification>> {
+		self.storage.justification(id)
+	}
+
+	fn info(&self) -> error::Result<blockchain::Info> {
+		self.storage.info()
+	}
+
+	fn status(&self, id: BlockId) -> error::Result<BlockStatus> {
+		self.storage.status(id)
+	}
+
+	fn hash(&self, number: block::Number) -> error::Result<Option<block::HeaderHash>> {
+		self.storage.hash(number)
+	}
+}
+
+impl StateBackend for OnDemandState {
+	type Error = error::Error;
+	type Transaction = ();
+
+	fn storage(&self, _key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+		// TODO [light]: fetch from remote node
+		Err(error::ErrorKind::NotAvailableOnLightClient.into())
```

Comment:
IMO this is really not the right approach. Making round trips for every storage entry we want to read <<<< making a single round trip for an entire call.

Reply:
There should be a `StateBackend` which is generated from an execution proof; then all the methods can be implemented.

Reply:
There's `impl StateBackend for OnDemandState` - stub right now. `storage` is not intended to be used during calls - as mentioned in description:

'light backend' should fetch required data from remote nodes using provided Fetcher. Since there's separate layer for executing calls (CallExecutor), it should only fetch a data for RPC calls (when properly used), but still it should be a fully-functional backend;